v2/comments.go

```diff
@@ -229,132 +241,206 @@ func (t Tag) String() string {
-	parts := strings.SplitN(input, "(", 2)
-	key := parts[0]
-
-	if len(tagNames) > 0 {
-		found := false
-		for _, tn := range tagNames {
-			if key == tn {
-				found = true
-				break
-			}
-		}
-		if !found {
-			return "", nil, nil
-		}
+func (s *tagKeyScanner) parseTagKey(tagNames []string) (string, []string, error) {
+	if s.Scan() != scanner.Ident {
+		return "", nil, fmt.Errorf("expected identifier but got %q", s.TokenText())
 	}
-
-	var args []string
-	if len(parts) == 2 {
-		if ret, err := parseTagArgs(parts[1]); err != nil {
-			return key, nil, fmt.Errorf("failed to parse tag args: %v", err)
-		} else {
-			args = ret
-		}
+	tagName := s.TokenText()
+	if len(tagNames) > 0 && !slices.Contains(tagNames, tagName) {
+		return "", nil, nil
+	}
+	if s.Peek() != '(' {
+		return tagName, nil, nil
+	}
+	s.Scan() // consume the '(' token
+	args, err := s.parseTagArgs()
+	if err != nil {
+		return "", nil, err
+	}
+	if s.Scan() != ')' {
+		return "", nil, s.unexpectedTokenError("')'", s.TokenText())
 	}
-	return key, args, nil
+	return tagName, args, nil
 }
 
 // parseTagArgs parses the arguments part of an extended comment tag. The input
 // is assumed to be the entire text of the original input after the opening
-// '(', including the trailing ')'.
+// '(', and before the trailing ')'.
 //
-// At the moment this assumes that the entire string between the opening '('
-// and the trailing ')' is a single Go-style identifier token OR a raw string
-// literal. The single Go-style token may consist only of letters and digits
-// and whitespace is not allowed.
-func parseTagArgs(input string) ([]string, error) {
-	s := initArgScanner(input)
-	var args []string
-	if s.Peek() != ')' {
-		// Arg found.
-		arg, err := parseArg(s)
+// The argument may be a go style identifier, a quoted string ("..."), or a raw string (`...`).
+func (s *tagKeyScanner) parseTagArgs() ([]string, error) {
+	if s.Peek() == ')' {
+		return nil, nil
+	}
+	if s.Peek() == '{' || s.Peek() == '[' {
+		value, err := s.scanJSONFlavoredValue()
 		if err != nil {
 			return nil, err
 		}
-		args = append(args, arg)
+		return []string{value}, nil
 	}
-	// Expect one closing ')' after the arg.
-	if s.Scan() != ')' {
-		return nil, fmt.Errorf("no closing ')' found: %q", input)
+	switch s.Scan() {
+	case scanner.String, scanner.RawString, scanner.Ident:
+		return []string{s.TokenText()}, nil
+	default:
+		return nil, s.unexpectedTokenError("identifier, quoted string (\"...\") or raw string (`...`)", s.TokenText())
 	}
-	// Expect no whitespace, etc. after the one ')'.
-	if s.Scan() != scanner.EOF {
-		pos := s.Pos().Offset - len(s.TokenText())
-		return nil, fmt.Errorf("unexpected characters after ')': %q", input[pos:])
+}
+
+// scanJSONFlavoredValue consumes a single token as a JSON value from the scanner and returns the token text.
+// A strict subset of JSON is supported, in particular:
+// - Big numbers and numbers with exponents are not supported.
```

Comment:
nit: Can we add tests for these cases so that the subset is confirmed and codified in tests?

Reply:
Added